src/dawn/IIR/StatementAccessesPairIterator.h

```diff
@@ -0,0 +1,131 @@
+﻿//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#ifndef DAWN_IIR_SAPITERATOR_H
+#define DAWN_IIR_SAPITERATOR_H
+
+#include "dawn/IIR/IIRNode.h"
+#include "dawn/IIR/IIRNodeIterator.h"
+#include "dawn/IIR/StatementAccessesPair.h"
+#include "dawn/SIR/ASTNodeIterator.h"
+#include "dawn/Support/IteratorRange.h"
+#include <unordered_map>
+
+namespace dawn {
+
+/// @brief Iterator through StatementAccessesPairs of statements of an AST
+///
+/// Built on top of an ASTNodeIterator that iterates through statements, requires a map from ASTStmt
+/// to StatementAccessesPair. Can either visit the first level of the AST or the whole AST.
+template <ASTNodeIteratorVisitKind onlyFirstLevel>
+class StatementAccessesPairIterator {
+public:
+  using WrappedIterator = ASTNodeIterator<onlyFirstLevel>;
+  using ASTStmtToSAPMapType = std::unordered_map<const Stmt*, iir::StatementAccessesPair>;
+
+  StatementAccessesPairIterator(const std::shared_ptr<Stmt>& root,
+                                const ASTStmtToSAPMapType& ASTStmtToSAPMap)
+      : current_(WrappedIterator::CreateInstance(root)),
+        ASTStmtToSAPMap_(const_cast<ASTStmtToSAPMapType&>(ASTStmtToSAPMap)) {}
+  StatementAccessesPairIterator(const WrappedIterator& it,
+                                const ASTStmtToSAPMapType& ASTStmtToSAPMap)
+      : current_(it.clone()), ASTStmtToSAPMap_(const_cast<ASTStmtToSAPMapType&>(ASTStmtToSAPMap)) {}
```

Comment:
I don't think we should ever `const_cast` when not dealing with legacy code or really difficult getters.

Reply:
Making `ASTStmtToSAPMap_` a `const` member of `StatementAccessesPairIterator` would make the call to `ASTStmtToSAPMap_.at(...)` return a  `const StatementAccessesPair&`: https://github.com/Stagno/dawn/blob/4765d92cb3c200e49ec8464d33d59492aac462b9/src/dawn/IIR/StatementAccessesPairIterator.h#L56
while we want to return a `StatementAccessesPair&`.